ecf/scripts/gdas/atmos/gempak/jgdas_atmos_gempak.ecf

```diff
@@ -35,6 +35,12 @@ export cyc=%CYC%
 export cycle=t%CYC%z
 export USE_CFP=YES
 
+#### EMC developer only
+COMPONENT=atmos
+export COMIN=${COMIN:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT}
+export COMOUT=${COMOUT:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT/gempak}
+####
+
```

Comment:
Why are these lines necessary? For this ecf script...the JJOB that gets run right after these lines already contains essentially the same exact lines:

https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/JGDAS_ATMOS_GEMPAK#L62:L68

The JJOB updates in the `feature/ops-wcoss2` need to come back to the `develop` branch still...perhaps those updates will remove the need for these lines? See:

https://github.com/NOAA-EMC/global-workflow/blob/feature/ops-wcoss2/jobs/JGDAS_ATMOS_GEMPAK#L82:L88

We can set developer COM* variables via `config.base.emc.dyn` also if that helps. 

Similar question for all of the other ecf scripts getting `COMPONENT`, `COMIN` and `COMOUT` paths set just for dev mode. Are they needed? What happens without them?

Reply:
@KateFriedman-NOAA it is not true from your statement - these lines already contains essentially the same exact lines
Here is the [JGDAS_ATMOS_GEMPAK#L62:L68](https://github.com/NOAA-EMC/global-workflow/blob/develop/jobs/JGDAS_ATMOS_GEMPAK#L62:L68) 
##############################################
# Define COM directories
##############################################
export COMIN=${COMIN:-$(compath.py ${NET}/${envir}/${RUN}.${PDY})/${cyc}/$COMPONENT}
export COMOUT=${COMOUT:-${COMROOT}/${NET}/${envir}/${RUN}.${PDY}/${cyc}/$COMPONENT/gempak}

It is obviously different compare to the solution provided:
export COMIN=${COMIN:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT}
export COMOUT=${COMOUT:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT/gempak}

The one without "-o" will result in NCO location. The one with "-o" can be assigned to a developer location. These answer will apply to all other similar question for ecf scripts getting COMPONENT, COMIN and COMOUT paths set just for dev mode.

Reply:
Did you look at the updated line in the feature/ops-wcoss2 branch? See:

https://github.com/NOAA-EMC/global-workflow/blob/feature/ops-wcoss2/jobs/JGDAS_ATMOS_GEMPAK#L82:L88

You're adding the following into the ecf script (for dev-mode only, can't use for ops):
```
COMPONENT=atmos
export COMIN=${COMIN:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT}
export COMOUT=${COMOUT:-$(compath.py -o ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT/gempak}
```
...and the updated script in `feature/ops-wcoss2` (ignore the `develop` branch copy, it will be replaced by the copy from `feature/ops-wcoss2` soon) has:
```
export COMPONENT=${COMPONENT:-atmos}
export COMIN=${COMIN:-$(compath.py ${envir}/${NET}/${gfs_ver})/${RUN}.${PDY}/${cyc}/$COMPONENT}
export COMOUT=${COMOUT:-$(compath.py -o ${NET}/${gfs_ver}/${RUN}.${PDY})/${cyc}/$COMPONENT/gempak}
```

The latter is vetted, tested, and in use in ops on WCOSS2. Other than hardwiring `COMPONENT` in the ecf script (which I don't see as necessary) the only differences are:
1) the `-o` flag with `COMIN` in the ecf script setting you added
2) setting `${envir}` in the ecf script setting you added

Why not update the JJOB script if you're saying both of those are needed? Would need to make sure those edits don't break ops of course.

Trying to avoid developer-only stuff in the ecf scripts that we would just have to strip out for ops anyway (not good).

Reply:
Would need to talk to NCO about `-o` with `COMIN` and `${envir}` with `COMOUT`...I remember there was a reason why they weren't included but it's not fresh in my memory.

Reply:
At the time when I test ecflow with Wei Wei, I used this solution to assign developer COM path. It was no red flag reported by NCO with this solution in development. Russ reported similar issue using path assignment without the -o option to @KateFriedman-NOAA during the wcoss2 testing. I communicated back to Russ and Kate provided this same workaround solution.
If NCO decided to apply a different solution, we can modify this solution as needed.

Reply:
happy to chat about this anytime. definitely don’t want -o for COMIN and definitely don’t want dev only blocks in ops. compath was designed to be flexible for ops and dev; if there’s a use case where that’s not true then we can always update compath to be more flexible, while avoiding dev blocks

Reply:
We can make any change needed.

Here is the original justification for using -o for developer COM path assignment:
Developer use customer configured path for development parallel runs, that should not have any wired production path. 
For example, if developer want to setup COM=/lfs/h2/emc/ptmp/$USER/$PSLOT/com/para/gfs/v16.2
Using compath.py para/gfs/v16.2.0 resulting:
/lfs/h1/ops/para/com/gfs/v16.2 (wired in production location)
Using the following -o solution will allow developer to get the requested COM path:
export COMROOT=/lfs/h2/emc/ptmp/$USER/$PSLOT/com
compath.py -o para/gfs/v16.2.0
/lfs/h2/emc/ptmp/$USER/$PSLOT/com/para/gfs/v16.2

Regarding generalize it, historically some jobs does not source $HOMEgfs/parm/config files. This condition also exist during wcoss2 transition. Some jobs using COMIN and COMOUT. However, some jobs using COMINgfs, COM_IN, COMINgempak, TANKverf ... Moving developer COM path assignment into include directory required unified COM path assignment from all jobs. Hence the reason of using developer block with -o on those ecflow script in question.

Reply:
A meeting on April 21th, 2022 with NCO. The requirement of this PR has been changed to:
- NCO will modify compath.py utility to allow developer to overwrite.
- EMC will modify ecflow script to remove EMC developer only area.
- EMC will encode new compath.py in envir-p1.h with generalized high level path assignment.
- EMC will modify the JJobs, and ex-scripts for these 25 jobs to use the new generalized high level path assignment.
- PR the modified 25 ecflow scripts and envir-p1.h after the above change is completed.

Reply:
compath.py for developers.... @lgannoaa @aerorahul 
Steven.Earle@clogin01:~>  module load prod_util
Steven.Earle@clogin01:~> export COMPATH=/lfs/h1/nco/idsb/noscrub/Steven.Earle/para/com/gfs
Steven.Earle@clogin01:~> compath.py gfs/v16.2
/lfs/h1/nco/idsb/noscrub/Steven.Earle/para/com/gfs/v16.2

Reply:
> compath.py for developers
@StevenEarle-NCO is this message indicated that the action item 1 - NCO will modify compath.py utility to allow developer to overwrite. is now completed? 
A test assume developer want to use /lfs/h2/emc/ptmp/$USER/$PSLOT/com/para/gfs/v16.2 as COM:
export COMPATH=/lfs/h2/emc/ptmp/Lin.Gan/ecfops/com/para/gfs     
Lin.Gan@clogin09:> compath.py gfs/v16.2
/lfs/h2/emc/ptmp/Lin.Gan/ecfops/com/gfs/v16.2
Lin.Gan@clogin09:> compath.py para/gfs/v16.2
/lfs/h2/emc/ptmp/Lin.Gan/ecfops/com/gfs/v16.2
Both of the above tries are unable to overwrite COM path with desired developer COM location.

Reply:
It looks to me like your testing worked... but you can't have /com/para/gfs as this is the wcoss1 structure. For wcoss2 structure, it will work to overwrite; please let me know if I'm missing something...
Change your development path to para/com/gfs/v16.2 -- it should work.

Steven